server/sonar-process-monitor/src/main/java/org/sonar/application/process/SQProcess.java

```diff
@@ -0,0 +1,262 @@
+  // keep flag so that the operational event is sent only once
+  // to listeners
+  private final AtomicBoolean operational = new AtomicBoolean(false);
+
+  private SQProcess(Builder builder) {
+    this.processId = requireNonNull(builder.processId, "processId can't be null");
+    this.lifecycle = new Lifecycle(this.processId, builder.lifecycleListeners);
+    this.eventListeners = builder.eventListeners;
+    this.watcherDelayMs = builder.watcherDelayMs;
+    this.stopWatcher = new StopWatcher();
+    this.eventWatcher = new EventWatcher();
+  }
+
+  public boolean start(Supplier<ProcessMonitor> commandLauncher) {
+    if (!lifecycle.tryToMoveTo(Lifecycle.State.STARTING)) {
+      // has already been started
+      return false;
+    }
+    try {
+      this.process = commandLauncher.get();
+    } catch (RuntimeException e) {
+      LOG.error(format("Fail to launch process [%s]", processId.getKey()), e);
+      lifecycle.tryToMoveTo(Lifecycle.State.STOPPED);
+      throw e;
+    }
+    this.gobbler = new StreamGobbler(process.getInputStream(), processId.getKey());
+    this.gobbler.start();
+    this.stopWatcher.start();
+    this.eventWatcher.start();
+    // Could be improved by checking the status "up" in shared memory.
+    // Not a problem so far as this state is not used by listeners.
+    lifecycle.tryToMoveTo(Lifecycle.State.STARTED);
+    return true;
+  }
+
+  public ProcessId getProcessId() {
+    return processId;
+  }
+
+  Lifecycle.State getState() {
+    return lifecycle.getState();
+  }
+
+  /**
+   * Sends kill signal and awaits termination. No guarantee that process is gracefully terminated (=shutdown hooks
+   * executed). It depends on OS.
+   */
+  public void stop(long timeout, TimeUnit timeoutUnit) {
+    if (lifecycle.tryToMoveTo(Lifecycle.State.STOPPING)) {
+      stopGracefully(timeout, timeoutUnit);
+      if (process != null && process.isAlive()) {
+        LOG.info("{} failed to stop in a timely fashion. Killing it.", processId.getKey());
+      }
+      // enforce stop and clean-up even if process has been gracefully stopped
+      stopForcibly();
+    } else {
+      // already stopping or stopped
+      waitForDown();
+    }
+  }
+
+  private void waitForDown() {
+    while (process != null && process.isAlive()) {
+      try {
+        process.waitFor();
+      } catch (InterruptedException ignored) {
+        // ignore, waiting for process to stop
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  private void stopGracefully(long timeout, TimeUnit timeoutUnit) {
+    if (process == null) {
+      return;
+    }
+    try {
+      // request graceful stop
+      process.askForStop();
+      process.waitFor(timeout, timeoutUnit);
+    } catch (InterruptedException e) {
+      // can't wait for the termination of process. Let's assume it's down.
+      LOG.warn(format("Interrupted while stopping process %s", processId), e);
+      Thread.currentThread().interrupt();
+    } catch (Throwable e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1181)